### PR TITLE
[WIP] isin method for columns and indexes

### PR DESCRIPTION
**Pull request by beckernick**
**Summary (updated)**
- This PR implements the `isin` method for Series of numeric/datetime/string types, and on the base Index class (though CategoricalIndexes will still error like CategoricalColumns for now). It matches pandas in general, but the typecasting behavior involved can cause some edge cases. Several edge cases are now xfailing pytests for future work.
- Nulls in the values array are handled **differently** than pandas. Nulls are equal to each other in pandas `isin`:

```python
data = pd.Series([50.0, 13, None])
values = [None, 13]
print(data.isin(values))
0    False
1     True
2     True
dtype: bool
```
Instead, we return :
```python
data = cudf.Series([50.0, 13, None])
values = [None, 13]
print(data.isin(values))
0    False
1     True
2    False
dtype: bool
```

and raise a warning to the user.

This functionality makes it possible to do things like the following in Python:

```python
import cudf
df = cudf.DataFrame({'a':[0,1,None,3], 'b':['a','b','c','d'], 'c':[4,3,2,1]})
print(df)
   a  b  c
0  0  a  4
1  1  b  3
2     c  2
3  3  d  1
print(df[df.a.isin(df.c)])
   a  b  c
1  1  b  3
3  3  d  1
```

Or to pass in CPU array at runtime:

```python
keepvals = [0,3]
print(df[df.a.isin(keepvals)])
   a  b  c
0  0  a  4
3  3  d  1
```
This PR depends on #2316

### CODE REVIEW

File: python/cudf/cudf/dataframe/datetime.py [hunk 1]
```diff
@@ -255,6 +257,24 @@ def is_monotonic_decreasing(self):
             ).all()
         return self._is_monotonic_decreasing
 
+    def isin(self, values):
+        from cudf.dataframe import numerical
+        mask = None
+
+        if isinstance(values, (list, np.ndarray,)):
+            values = DatetimeColumn.from_numpy(
+                np.asarray(values)).as_numerical.data.mem
```
Comment (kkraus14) [inline python/cudf/cudf/dataframe/datetime.py]: you can't assume the values are coming from the host side here. I'd use `columnops.as_column` instead.

File: python/cudf/cudf/dataframe/numerical.py [hunk 2]
```diff
@@ -283,6 +283,21 @@ def round(self, decimals=0):
             data=Buffer(rounded), mask=mask, dtype=self.dtype
         )
 
+    def isin(self, values):
+        mask = None
+
+        if isinstance(values, (list, np.ndarray,)):
+            values = rmm.to_device(np.asarray(values))
```
Comment (kkraus14) [inline python/cudf/cudf/dataframe/numerical.py]: same here

**Comment by beckernick** [comment]
@kkraus14 I agree that a hash table based approach would be superior as a stopgap until libcudf support as it would not also scale linearly (worse case) in `values` array axis, beyond the scaling in the `nrows` axis. Implementing this with join(s) felt more complex as I think it might require a bit more tricky logic than seems like at first, but after doing a bit more nuanced profiling it seems like that may be the right choice.

I ran some initial time tests and appear to have been a bit mislead by the times (or something odd is going on, but I'd bet the former). The actual time spent in executing `isin` seemed to be quite low at scale (0.16 seconds, for example). But, lots of time is spent in Numba/RMM related memory management (if the data is a host array) or in cuDF related indexing (if the data is already a Series) in operations seemingly _after_ the execution of the kernel. I'm not familiar enough with Numba to know exactly what's occurring here, yet. Either way, the net net is that this is slower at scale than I initially realized since we'd want to do things after the kernel executes.

If I'm not just missing something obvious, it may be worth us looking into what's happening here (likely the issue is just my understanding, though).

Profile of isin when the values array is on the host (numpy):
```python
import cudf
import numpy as np
import time

nrows = 1000000
df = cudf.DataFrame()
df['a'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
df['b'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
df['c'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
keepvals = np.random.randint(-1000, 100000, 1000000)

t00 = time.time()

t0 = time.time()
res = df.a.isin(keepvals)
t1 = time.time()
print('isin time keepvals: {}'.format(t1-t0))
print('done')
res.head()

t11 = time.time()
print('total time: {}'.format(t11 - t00))
isin time keepvals: 5.436095237731934
done
total time: 5.525848627090454
```
Screen Shot 2019-07-16 at 11 02 42 AM

Note the 0.17 seconds in `numerical.py` is the path in which `apply_isin` is executed (pink rectangle).


Profile of isin when the values array is another cudf series:
Screen Shot 2019-07-16 at 11 12 12 AM

```python
import cudf
import numpy as np
import time
nrows = 1000000
df = cudf.DataFrame()
df['a'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
df['b'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
df['c'] = cudf.Series(np.random.randint(-100000, 100000, nrows))
keepvals = np.random.randint(-1000, 100000, 1000000)

t00 = time.time()

t0 = time.time()
res = df.a.isin(df.b)
t1 = time.time()
print('isin time df.b: {}'.format(t1-t0))
print('done')
res.head()

t11 = time.time()
print('total time: {}'.format(t11 - t00))
```
Again, just 0.17 seconds in codepath of `series.isin` (pink rectangle). The profile makes it seem like the time is spent within the call to `series.head`.

**Comment by trxcllnt** [comment]
I pushed an update to the tests and an implementation that's compatible with Categoricals. This new version also updates `as_column` to cast arbitrary python lists to nvstrings if the dtype is `"object"`.

**Comment by harrism** [comment]
I think this PR is mal-formed -- look at the commit history, it goes back to April. :)  I think you need to merge an up-to-date branch-0.9 into your feature branch.

**Comment by trxcllnt** [comment]
fyi this method is also in PR https://github.com/rapidsai/cudf/pull/2382. I needed it for `cat.remove_categories()`, so @beckernick and I worked together on this earlier this week, then I cherry-picked the commits. If fixing the history in this PR is too much of a headache, we can either make a fresh PR from the other branch, or just wait till #2382 is merged.

**Comment by harrism** [comment]
Please mark this issue either "in progress" or "ready for review" so we know if it's waiting on a review. Not really reviewable until the commit history is cleaned up, so I assume in progress.

**Comment by beckernick** [comment]
Given the commit history blowup, let's just use the other PR @trxcllnt . As long as review/merge of that PR is not blocked by any outstanding issues, no harm done. 

This PR has been subsumed under #2382 . Closing.

**Pull request closed without merging**